stan/math/prim/mat/fun/value_of_rec.hpp

```diff
@@ -22,12 +22,33 @@ namespace stan {
     template <typename T, int R, int C>
     inline Eigen::Matrix<double, R, C>
     value_of_rec(const Eigen::Matrix<T, R, C>& M) {
-      using stan::math::value_of_rec;
-      Eigen::Matrix<double, R, C> Md(M.rows(), M.cols());
-      for (int j = 0; j < M.cols(); j++)
-        for (int i = 0; i < M.rows(); i++)
-          Md(i, j) = value_of_rec(M(i, j));
-      return Md;
+      Eigen::Matrix<double, R, C>
+        result(M.rows(), M.cols());
+
+      double * data_r = result.data();
+      const T * data_m = M.data();
+
+      int S = M.size();
+      for (int i=0; i < S; i++)
+        data_r[i] = value_of_rec(data_m[i]);
+      return result;
+    }
```

Comment:
Also, with -O3, this

       for (size_t i = 0; i < xs.rows(); ++i)
       for (size_t j = 0; j < xs.cols(); ++j)

doesn't seem to be slower than this:

       for (size_t i = 0, nRows = xs.rows(), nCols = xs.cols(); i < 
nRows; ++i)
       for (size_t j = 0; j < nCols; ++j)

even when xs is not const.

(Sorry if it sound rather obvious that the compiler's able to optimize this)